pkg/rbd/controllerserver.go

```diff
@@ -391,18 +391,24 @@ func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
 
 	// Deleting rbd image
 	klog.V(4).Infof(util.Log(ctx, "deleting image %s"), rbdVol.RbdImageName)
-	if err := deleteImage(ctx, rbdVol, cr); err != nil {
+	if err = deleteImage(ctx, rbdVol, cr); err != nil {
 		klog.Errorf(util.Log(ctx, "failed to delete rbd image: %s/%s with error: %v"),
 			rbdVol.Pool, rbdVol.RbdImageName, err)
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 
-	if err := undoVolReservation(ctx, rbdVol, cr); err != nil {
+	if err = undoVolReservation(ctx, rbdVol, cr); err != nil {
 		klog.Errorf(util.Log(ctx, "failed to remove reservation for volume (%s) with backing image (%s) (%s)"),
 			rbdVol.RequestName, rbdVol.RbdImageName, err)
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 
+	if rbdVol.Encrypted {
+		if err = rbdVol.KMS.DeletePassphrase(rbdVol.VolID); err != nil {
```

Comment:
@Madhu-1 Note for PR https://github.com/ceph/ceph-csi/pull/693 : 
- Currently the base image can only be deleted once all the snapshots and clones are deleted, afaik
- Hence deleting this KMS key when the base image is deleted should be fine, as all dependent snaps and clones using this key would already be deleted
- Once the said PR is implemented though, we should be able to delete base image indenpendent of other relations (although this just means moving it to trash), so the key would also be deleted. This may need some handling, as we do not want to lose a key that has data in other related images encrypted.

(although in the final solution, we do not want snapshots and clones to share the same key, but till then we need the snapshot/clone story with encryption clear)

Reply:
I think more secure approach would be to prevent creating snapshots from encrypted volumes. As mentioned before - creating many copies that use same encryption key opens it to attacks. Also I believe this will depend heavily on implementing re-encryption for volumes.

Reply:
> I think more secure approach would be to prevent creating snapshots from encrypted volumes. As mentioned before - creating many copies that use same encryption key opens it to attacks. Also I believe this will depend heavily on implementing re-encryption for volumes.

Ack! so no intermediate solutions, we will turn off snapshots/clones for encrypted volumes, till we add offline re-encrypt support. (@Madhu-1 something for PR #693 I would assume)

Reply:
add guard for this [here](https://github.com/ceph/ceph-csi/pull/762/files#diff-35c77cc181d49675cd67a6095520b41fR465-R468)